aten/src/ATen/native/UpSampleNearest1d.cpp

```diff
@@ -0,0 +1,207 @@
+      pos2 += output_width;
+    }
+  }
+}
+
+static void upsample_nearest1d_out_cpu_template(
+    Tensor& output,
+    const Tensor& input_,
+    IntArrayRef output_size) {
+  int64_t output_width = output_size[0];
+
+  int64_t nbatch = input_.size(0);
+  int64_t channels = input_.size(1);
+  int64_t input_width = input_.size(2);
+
+  upsample_1d_shape_check(
+      input_,
+      static_cast<int64_t>(0),
+      nbatch,
+      channels,
+      input_width,
+      output_width);
+
+  auto input = input_.contiguous();
+
+  output.resize_({nbatch, channels, output_width});
+  output.zero_();
+
+  AT_ASSERT(input_width > 0 && output_width > 0);
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "upsample_nearest1d", [&] {
+    auto* idata = input.data<scalar_t>();
+    auto* odata = output.data<scalar_t>();
+
+    upsample_nearest1d_out_frame<scalar_t>(
+        odata,
+        idata,
+        input_width,
+        output_width,
+        nbatch,
+        channels);
+  });
+}
+
+static void upsample_nearest1d_backward_out_cpu_template(
+    Tensor& grad_input,
+    const Tensor& grad_output_,
+    IntArrayRef output_size,
+    IntArrayRef input_size) {
+  int64_t output_width = output_size[0];
+
+  int64_t nbatch = input_size[0];
+  int64_t channels = input_size[1];
+  int64_t input_width = input_size[2];
+
+  upsample_1d_shape_check(
+      grad_output_,
+      static_cast<int64_t>(1),
+      nbatch,
+      channels,
+      input_width,
+      output_width);
+
+  auto grad_output = grad_output_.contiguous();
+
+  grad_input.resize_({nbatch, channels, input_width});
+  grad_input.zero_();
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      grad_output.scalar_type(), "upsample_nearest1d_backward", [&] {
+        scalar_t* idata = grad_input.data<scalar_t>();
+        scalar_t* odata = grad_output.data<scalar_t>();
+
+        upsample_nearest1d_backward_out_frame<scalar_t>(
+            odata,
+            idata,
+            input_width,
+            output_width,
+            nbatch,
+            channels);
+      });
+}
+} // namespace
+
+Tensor& upsample_nearest1d_out_cpu(
+    Tensor& output,
+    const Tensor& input,
+    IntArrayRef output_size) {
+  upsample_nearest1d_out_cpu_template(output, input, output_size);
+  return output;
+}
+
+Tensor upsample_nearest1d_cpu(const Tensor& input, IntArrayRef output_size) {
+  auto output = at::empty({0}, input.options());
+  upsample_nearest1d_out_cpu_template(output, input, output_size);
+  return output;
+}
+
+Tensor& upsample_nearest1d_backward_out_cpu(
+    Tensor& grad_input,
+    const Tensor& grad_output,
+    IntArrayRef output_size,
+    IntArrayRef input_size) {
+  upsample_nearest1d_backward_out_cpu_template(
+      grad_input, grad_output, output_size, input_size);
+  return grad_input;
+}
+
+Tensor upsample_nearest1d_backward_cpu(
+    const Tensor& grad_output,
+    IntArrayRef output_size,
+    IntArrayRef input_size) {
+  auto grad_input = at::zeros_like(grad_output);
```

Comment:
This looks wrong. It should be zeros with the input size. `x` and `grad_x` have the same size.

Reply:
would `auto grad_input = at::zeros(input_size, grad_output.options());` be correct here?